File: solutions/clojure/bob/1/src/bob.clj
(ns bob)

(require '[clojure.string :as str])

(defn question? [s]
  (= (last s) \?))

(defn yelling? [s]
  (and (re-find #"[A-Z]" s)
       (not (re-find #"[a-z]" s))))

(defn silence? [s]
  (= (str/trim s) ""))

(defn response-for [s]
  (let [s (str/trim s)]
    (cond
      (and (question? s) (yelling? s)) "Calm down, I know what I'm doing!"
      (question? s) "Sure."
      (yelling? s) "Whoa, chill out!"
      (silence? s) "Fine. Be that way!"
      :else "Whatever.")))

File: solutions/haskell/bob/1/src/Bob.hs
module Bob (responseFor) where
-- import Distribution.Simple.Utils (xargs)
import Data.Char (isSpace)

responseFor :: String -> String
responseFor xs
    | null str = "Fine. Be that way!"
    | isQuestion && isYell = "Calm down, I know what I'm doing!"
    | isQuestion = "Sure."
    | isYell = "Whoa, chill out!"
    | otherwise = "Whatever."
    where str = filter (not . isSpace) xs
          isQuestion = last str == '?'
          isYell = hasUpper && not hasLower
              where hasUpper = any (`elem` ['A'..'Z']) str
                    hasLower = any (`elem` ['a'..'z']) str
